hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+
+When we consider how hash tables are designed and implemented, we will:
+
+- Practice algorithmic and design thinking around data structures
+- Consider how complex data structures are implemented
+- Learn powerful problem-solving techniques for problems often encountered in interviews
+- Appreciate dictionaries more!
+
+## Vocabulary and Synonyms
+
+| Vocab         | Definition                                                                                          | Synonyms         | How to Use in a Sentence                                                                                                                                                                                      |
+| ------------- | --------------------------------------------------------------------------------------------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Hash Table    | A data structure that implements an associative array, which maps unique key identifiers to values. | Hash Map         | "Hash tables have an O(1) lookup time."                                                                                                                                                                       |
+| Hash Function | A function used by a hash table to map a key to an index.                                           | Hashing function | "The hash function converts a key into an index in an associative array"                                                                                                                                      |
+| Collision     | When multiple keys map to the same element in a hash table's internal array.                        | -                | "The hash function wanted to insert a value into an index where a different value already existed, which is a collision. The hash table needs to handle this collision in order to keep both pieces of data." |
+
+# Hash Tables
+
+A _hash table_ is a data structure designed to store key-value pairs. They have a method to look up a value by its key (find), add a key-value pair (insert), and delete a key-value pair (remove).
+
+Hash tables can uniquely look up values by key with a time complexity of O(1). With our current knowledge of programming, how is it possible for a data structure to achieve this, even though searching for an item in an array using linear search has a time complexity of O(n)?
+
+Hash tables achieve their speedy lookup times of O(1) by storing their key-value pairs into an array. They take advantage of how arrays have O(1) lookup operations when given an index.
+
+Each hash table has an _internal array_ used for storing. When a hash table inserts a key-value pair, it stores the value into the _internal array_.
```

Comment:
the entire pair gets stored in the internal array. this is necessary so that after finding the correct index for some data, the hash table still needs to perform a key equality comparison (since there could be collisions)
```suggestion
Each hash table has an _internal array_ that it uses for its storage. When a hash table inserts a key-value pair, it stores the pair in the _internal array_ at a particular position.
```